src/runtime/formal/ops/ops.cc

```diff
@@ -227,47 +203,40 @@ CVM_REGISTER_GLOBAL("cvm.runtime.formal.max_pool2d")
   int stride_h = param.strides[0];
   int stride_w = param.strides[1];
 
-  int32_t* x_data = (int32_t*)x->data;
-  int32_t* y_data = (int32_t*)y->data;
-
   int filter_h = param.pool_size[0];
   int filter_w = param.pool_size[1];
 
-  int n_batch = static_cast<int>(x->shape[0]);
-  int in_channels = static_cast<int>(x->shape[1]);
+  int n_batch = static_cast<int>(xShape[0]);
+  int in_channels = static_cast<int>(xShape[1]);
   int out_channels = in_channels;
-  int x_h = static_cast<int>(x->shape[2]);
-  int x_w = static_cast<int>(x->shape[3]);
-  int o_h = static_cast<int>(y->shape[2]);
-  int o_w = static_cast<int>(y->shape[3]);
-#define GETX(n, c, h, w) x_data[(n) * in_channels * x_h * x_w + (c) * x_h * x_w + (h) * x_w + (w)]
-#define GETY(n, c, h, w) y_data[(n) * out_channels * o_h * o_w + (c) * o_h * o_w + (h) * o_w + (w)]
-  auto calc_func = [&](int n, int k, int p, int q) {
+  int x_h = static_cast<int>(xShape[2]);
+  int x_w = static_cast<int>(xShape[3]);
+  int o_h = static_cast<int>(yShape[2]);
+  int o_w = static_cast<int>(yShape[3]);
+  auto calc_func = [&](const Indices& idx) {
+    int n = idx[0], k = idx[1], p = idx[2], q = idx[3];
+    // 1 << 31 will be the min value for int32: 0x80000000
     const int32_t minV = int32_t(1) << 31;
     int32_t y_max = minV;
+    Indices xIdx(xShape);
     for (int r = 0; r < filter_h; ++r) {
       for (int s = 0; s < filter_w; ++s) {
         int32_t tp = p * stride_h + r - padding[0];
         int32_t tq = q * stride_w + s - padding[1];
-        int32_t x_tmp = minV; 
-        if (0 <= tp && tp < x_h && 0 <= tq && tq < x_w)
-          x_tmp = GETX(n, k, tp, tq);
+        int32_t x_tmp = minV;
+        if (0 <= tp && tp < x_h && 0 <= tq && tq < x_w) {
+          xIdx.CopyIndicesFrom({idx[0], idx[1], tp, tq});
+          //std::cout << "xIndex is " << xIdx.Index() << std::endl;
+          x_tmp = x_data[xIdx.Index()];
+        }
         y_max = std::max(x_tmp, y_max);
```

Comment:
The `minV` variable is not necessary since you can move the assignment for `y_max` into
`if` block statements, whereas the outside bounding value keeps the same.